src/fpm_targets.f90

```diff
@@ -249,30 +274,87 @@ subroutine build_target_list(targets,model)
+        !
+        !
+        type(srcfile_t), intent(in) :: source
+        type(profile_config_t), intent(in) :: profile
+
+        character(:), allocatable :: file_scope_flag, current
+        integer :: i
+
+        file_scope_flag = ""
+
+        if (allocated(profile%file_scope_flags)) then
+            associate(fflags=>profile%file_scope_flags)
+                do i=1,size(fflags)
+                    if (source%file_name.eq.fflags(i)%file_name) then
+                        file_scope_flag = fflags(i)%flags//" "
+                    end if
+                end do
+            end associate
+        end if
+    end function get_file_scope_flags
+
     function get_object_name(source) result(object_file)
         ! Generate object target path from source name and model params
         !
         !
-        type(srcfile_t), intent(in) :: source
+        type(srcfile_t), intent(inout) :: source
         character(:), allocatable :: object_file
 
         integer :: i
         character(1), parameter :: filesep = '/'
-        character(:), allocatable :: dir
+        character(:), allocatable :: dir, out_dir, module_flags
 
         object_file = canon_path(source%file_name)
 
+        out_dir = get_output_directory(source)
+
+        call get_module_flags(model%fortran_compiler, out_dir, module_flags)
+
+        if (allocated(source%flags)) then
+            source%flags = source%flags // module_flags
+        else if (allocated(source%c_flags)) then
+            source%c_flags = source%c_flags // module_flags
+        end if
+
         ! Convert any remaining directory separators to underscores
         i = index(object_file,filesep)
         do while(i > 0)
             object_file(i:i) = '_'
             i = index(object_file,filesep)
         end do
 
-        object_file = join_path(model%output_directory,model%package_name,object_file)//'.o'
+        object_file = join_path(out_dir,model%package_name, object_file)//'.o'
 
     end function get_object_name
 
+    function get_output_directory(source) result(out_dir)
+        ! Generate build directory name by hashing the flags of the source
+        !
+        !
+        type(srcfile_t), intent(in) :: source
+
+        character(len=16) :: build_name
+        character(:), allocatable :: out_dir
+        type(string_t) :: include_dir
+
+        if (allocated(source%flags)) then
+            write(build_name, '(z16.16)') fnv_1a(source%flags)
+        else if (allocated(source%c_flags)) then
+            write(build_name, '(z16.16)') fnv_1a(source%c_flags)
+        end if
+        out_dir = join_path('build',basename(model%fortran_compiler)//'_'//build_name)
+        include_dir = string_t(out_dir)
+        if (.not. allocated(model%include_dirs)) then
+            model%include_dirs = [include_dir]
+        else if (.not. (out_dir.in.model%include_dirs)) then
+            model%include_dirs = [model%include_dirs, include_dir]
+        end if
```

Comment:
The model should not be modified once we reach `build_target_list`. For this logic, I suggest building a separate list of `include_dirs` from the output directories and then passing this separate list to `resolve_target_linking` in addition to `model%include_dirs`. To explain, `model%include_dirs` is independent of how we arrange our output objects whereas the output directories are specific to this particular layout of output objects. Hence `model%include_dirs` should be separate and immutable (in this module) to the backend-specific information. Let me know if that doesn't make sense.